### DISCUSSION TITLE
How to return to previous activities in c# code?

### DISCUSSION THREAD

**Discussion by hongjun-lee** [elsa-workflows/elsa-core#446] [Q&A]
In the Workflow,I Use two UserTask Activity, Activity A and Activity B. 
Suppose Activity B is executed,How can I return from B to A in C# code? I 'm not talking about Reject. So I can't use branch functions in design.
Please give me some guidance,thanks.

**Comment by sfmskywalker** [comment]
You can connect Activity B to Activity A in code by making sure Activity A has its `Name` property set to some value, e.g. `"ActivityA"`. You can then connect the outcome of Activity B to Activity A by specifying its name. For example:

```csharp
activityB
   .When("Still no good")
   .Then("ActivityA");
```

Or in full context based off Sample12:

```csharp
public class UserTaskWorkflow : IWorkflow
{
    public void Build(IWorkflowBuilder builder)
    {
        builder
            .StartWith<WriteLine>(activity => activity.TextExpression = new LiteralExpression("Workflow started. Waiting for user action."))
            .Then<UserTask>(
                activity => activity.Actions = new[] { "Accept", "Reject" },
                activityA =>
                {
                    activityA
                        .When("Accept")
                        .Then<WriteLine>(activity => activity.TextExpression = new LiteralExpression("Great! Your work has been accepted. Let's get this signed-off."))
                        .Then<UserTask>(activity => activity.Actions = new[]{ "Sign Off", "Request Changes" },
                            activityB =>
                            {
                                activityB
                                    .When("Sign Off")
                                    
                                    // All set. Connect to the Exit activity.
                                    .Then("Exit");
                                
                                activityB
                                    .When("Request Changes")
                                    
                                    // We need changes. Connect to the first UserTask activity (ActivityA).
                                    .Then("ActivityA");
                            });

                    activityA.When("Reject")
                        .Then<WriteLine>(activity => activity.TextExpression = new LiteralExpression("Sorry! Your work has been rejected."))
                        .Then("Exit");
                },
                "ActivityA"
            )
            .Add<WriteLine>(activity => activity.TextExpression = new LiteralExpression("Workflow finished."), "Exit");
    }
}
```

**Comment by hongjun-lee** [comment]
Thank you for your reply. 
My business scenario is to execute a "go back". Should I remember the source activity at the current node ActiveB and then arrange to return it to ActiveA.
Where can I remember the previous activities better? Remember it in Active's IWorkflowEventHandler.ActivityExecutedAsync()? 
If "go back" is executed, will it be recorded in ExecutionLog?

**Comment by hongjun-lee** [comment]
Add a few points:
This implementation of sample12 can work.
But if I use Workflow Design,How to achieve this jump in UserTask?
I'm now adding "Outcomes = "x => x.state.actions" to UserTask Activity.
The actions has {accept "," reject "}, but I don't think there should be "GoBack" because "GoBack" is dynamic and depends on where it comes from.
If ActiveA1 and ActivityA2 are connected to ActiveB, I need to know which branch runs to ActiveB, if it is A1, then go back to A1, if it is A2, the then Go back ot A2.
Sample12 is implemented in workflow builder, if we use the workflow definition of designer. How should we present "goback"?

**Comment by hongjun-lee** [comment]
Is it okay to use WorkflowExecutionContext.ScheduleActivity(IActivity activity)?

**Comment by sfmskywalker** [comment]
Ah I see. This is an interesting scenario!

Regarding the workflow designer, connecting an outcome of ActivityB to ActivityA1 or ActivityA2 is easy, you can just drag the correct endpoint from ActivityB to either ActivityA1 of ActivityA2.

But that doesn't provide you with the behavior you're after, which is literally a "go back" behavior as you mentioned.
There is nothing to facilitate this out of the box, but you can easily implement this yourself by writing a custom activity (could be a copy of the UserTask activity with modifications).

The custom activity would implement the logic to determine whether execution should go back to the previous activity.
Once that is determined, you need to determine what exactly the previous activity was. This can be done easily by getting the last `LogEntry` object from the workflow execution's `LogEntries` collection.

A `LogEntry` contains the activity ID that was executed. With this activity ID in hand, you now need to select the activity by ID (`var previousActivity = workflowContext.Workflow.GetActivity(thePreviousActivityId)l`), and then schedule it for execution like so:

`workflowContext.ScheduleActivity(previousActivity);`.

This will cause the previous activity to be executed again.

> If "go back" is executed, will it be recorded in ExecutionLog?
Yes, the previous activity will be recorded again in ExecutionLog.

If this works for your scenario, then I'm considering adding a new `ActivityExecutionResult` of type `GoBack` which automatically schedules the previously executed activity for you. This way, you only need to perform the logic to determine **if** you need to go back, and if so, simply return a `new GoBack()` result.

**Comment by sfmskywalker** [comment]
> Is it okay to use WorkflowExecutionContext.ScheduleActivity(IActivity activity)?

Yep, that's exactly what you should do when you get your hands on the previous activity.

**Comment by hongjun-lee** [comment]
Thank you very much. 
I look forward to your early realization of ActivityExecutionResult.
In addition, have you considered designing the tasklist function associated with actors? In the business scenario, after we associate the business with the workflow, we need to remind the actors on the current activity that they have work tasks to process.

**Comment by sfmskywalker** [comment]
I implemented it here (Elsa 2): 70d3569343b9b5fcf93be414e23cca7a994018b6
Do you need it for Elsa 1 as well?

**Comment by sfmskywalker** [comment]
Regarding the task list function associated with actors, this sounds application-specific. Can you perhaps elaborate a bit on what you would need in Elsa to achieve implementing such a workflow?

**Comment by hongjun-lee** [comment]
> I implemented it here (Elsa 2): [70d3569](https://github.com/elsa-workflows/elsa-core/commit/70d3569343b9b5fcf93be414e23cca7a994018b6)
> Do you need it for Elsa 1 as well?
Yes,I'm using elsa1 ,because I didn't run elsa2. I didn't see the manual. How do I run esla2?

**Comment by sfmskywalker** [comment]
Elsa 2 is currently in preview (no designer yet, so only works with programmatic workflows implemented as .NET classes). If you are interested you can try out [the preview bits](https://www.nuget.org/packages/Elsa/2.0.0-preview5-0).

There is no documentation as of yet, but there are close to 20 [examples](https://github.com/elsa-workflows/elsa-core/tree/feature/elsa-2.0/src/samples) to see how it works.

In any case, I'll back-port the GoBack result for Elsa 1 as well.

**Comment by sfmskywalker** [comment]
It's now part of Elsa 1.4.4.

**Comment by hongjun-lee** [comment]
My business scenario is like this. Multiple User Or Roles are involved in the Approval (UserTask) activity, so when the workflow executes this step, we need to notify the participants involved and tell them of the work to be done. My current implementation is to add a TaskList Table, and to insert data into TaskList when executing the Approval node.

![image](https://user-images.githubusercontent.com/73320360/97985015-8f753580-1e12-11eb-97e3-d9f64ef206a4.png)
So, I still have a few questions. please help me see if my realization idea is correct?

1. The Users And Roles involved in it, I define an Actor class and save the instance object in the Activity's State, because I don't want to destroy your existing program architecture, so that I can upgrade my program as you develop.

2. Because each activity involves multiple people (such as multi-player signatures), I want to remember each of their actions, so I want to define a 'ProcessLog' type to remember it. And keep 'StepInfo' in the State of each instance activity.

The code is roughly as follows: I'm writing this part. I hope I don't go around.   
![image](https://user-images.githubusercontent.com/73320360/97982752-32c44b80-1e0f-11eb-9592-d8a0f8c03589.png)

**Comment by hongjun-lee** [comment]
> It's now part of Elsa 1.4.4.
It's a great job. You are a good man. ^-^

**Comment by sfmskywalker** [comment]
I think I understand. You have a set of custom activities (e.g. `Initiator`, `DepartmentManagerApproval` and `AdminApproval`). Each of these activities record an actor's decision in the `ProcessLog` object.

I think I need to come across a couple more such business cases to try and distill the functionality that could be provided by Elsa that is general enough to support workflows such the one you showed, without being too opinionated.

In any case, I will keep this in the back of my mind.
Thank you for detailing out your use case, and for being awesome. ^-^

I'm closing this issue as the original question has been resolved. But feel free to open a new issue.

**Comment by hongjun-lee** [comment]
Okay, I'll try to write code to achieve my goal first. I'll consult you whenever I have any problems.

**Reply by Nokecy** [reply]
@hongjun-lee  Can you share part of your code

**Comment by sfmskywalker** [comment]
Sounds good.

**Reply by Nokecy** [reply]
@sfmskywalker  v2.8.x not found GoBackResult
